### ISSUE TITLE
zfs module doesn't support management of OpenZFS pools

### ISSUE THREAD

**Issue by xen0l** [ansible/ansible-modules-extras#2415]
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

zfs
##### ANSIBLE VERSION

ansible 2.1.0.0
##### CONFIGURATION

Nothing
##### OS / ENVIRONMENT

Freshly installed latest version of OpenZFS on OpenIndiana. This also affects every illumos-based operating systems (SmartOS, OmniOS etc)
##### SUMMARY

The problem is with zpool version. OpenZFS uses feature flags and sets zpool version to "-", which breaks compatibility with this module. It fails when it tries to compare OpenZFS zpool version with zpool version 34 on illumos-based system (https://github.com/ansible/ansible-modules-extras/blob/2a0c5e2a8fd7ed3ce6d6eedd08e85e01e1617113/system/zfs.py#L99)

This module should check for Solaris presence not by os.uname, but by ansible facts if possible.  
##### STEPS TO REPRODUCE

ansible all -m zfs -a 'name=rpool/data state=present' -vvv
##### EXPECTED RESULTS

ZFS dataset rpool/data should be created.
##### ACTUAL RESULTS

rpool/data wasn't created.

```
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_Y3bKol/ansible_module_zfs.py", line 245, in <module>
    main()
  File "/tmp/ansible_Y3bKol/ansible_module_zfs.py", line 227, in main
    zfs = Zfs(module, name, properties)
  File "/tmp/ansible_Y3bKol/ansible_module_zfs.py", line 90, in __init__
    self.enhanced_sharing = self.check_enhanced_sharing()
  File "/tmp/ansible_Y3bKol/ansible_module_zfs.py", line 99, in check_enhanced_sharing
    if int(version) >= 34:
ValueError: invalid literal for int() with base 10: '-'
```

**Comment by xen0l** [comment]
ping @johanwiren

**Comment by bcoca** [comment]
the module was working with BSD and linux, last I checked, so this is not only a Solaris issue

**Comment by johanwiren** [comment]
Yes, the OpenSolaris distributions have not been tested during development. Detecting them and handling them correctly should be easy to implement.

How to identify OpenSolaris distributions: https://gist.github.com/natefoo/7af6f3d47bb008669467

**Comment by xen0l** [comment]
@bcoca: do you know which commit broke the BSD and Linux support?

**Comment by xen0l** [comment]
@johanwiren: uname way doesn't apply to illumos distributions any longer. I am also thinking about adding a function, check_is_solaris, in which I open /etc/release and check for Oracle Solaris or Solaris (for Solaris 10) and set set.is_solaris. In the probmelatic fuction, check_enhanced_sharing, I shall decide based on self.is_solaris. What do you think?

**Comment by bcoca** [comment]
@xen0l, sorry I was unclear, I was not saying it is broken on those platforms, I was trying to indicate that a fix that changes version detection to 'Solaris' detection is not a viable solution.

**Comment by xen0l** [comment]
@bcoca what's the preferred fix then?

**Comment by bcoca** [comment]
I'm not sure, I don't know what version detection is controlling nor how/what interactions would change with openzfs, just that detecting 'solaris' is not a fix.

**Comment by xen0l** [comment]
Well, the Oracle ZFS has different feature set than OpenZFS. Oracle ZFS can run only on Oracle Solaris and detecting Oracle Solaris would help here. I'm afraid it's only viable solution. If the module stays general and doesn't harcode Oracle ZFS-centric features, we are good enough. Another option might be new module, but that will duplicate functionality. I'd take the solution I described earlier.

**Comment by bcoca** [comment]
That does not seem to be a correct assumption, as this module successfully ran on Linux and BSDs which are not using the Oracle ZFS implementation nor OpenZFS.

**Comment by xen0l** [comment]
@bcoca FreeBSD and Linux uses OpenZFS. Developers from those platforms collaborate on OpenZFS. OpenZFS just has illumos distributions as upstream. I found another solution and that is that OpenZFS reports different version "-" from Oracle Solaris. So, the question here is about not detecting Solaris, but detecting OpenZFS.

**Comment by bcoca** [comment]
FreeBSD had it's own implementation until recently, not all FreeBSD users run the latest, also it is not the only BSD, the same goes for linux, it has MANY implementations and many of them are in production.

OpenZFS is a relatively new consolidation of the efforts, I'm all for  supporting it, but not for sacrificing existing installed systems in the process.

**Comment by xen0l** [comment]
The change I am talking about will not affect those installations. The problem here is with detecting enhanced sharing properties, which is available since version 34 and applies only to Oracle Solaris. So, if not detecting OpenZFS vs Oracle ZFS, then detecting Solaris is the only way how to enable support on illumos platforms. So, are you OK with such solution then? If not, can you advise better approach?

**Comment by bcoca** [comment]
that does not sound right as the other implementations we discussed above would fall on the current behavior but you are detecting specifically against 'Solaris' to avoid it. I suggest detecing 'OpenZFS' instead as it seems to be the outlier in this case.

**Comment by olbohlen** [comment]
a very quick hack of course is in zfs.py: 

```
        version = out.splitlines()[-1].split()[2]
        if version == '-':
           return False
        if int(version) >= 34:
            return True
    return False
```

this will bail out if the pool version is matched to a string "-"...

**Comment by xen0l** [comment]
The fix is located at https://github.com/ansible/ansible-modules-extras/pull/2642. OpenZFS used to use version 5000 and uses "-" now. I added both checks.

**Comment by xen0l** [comment]
ping @johanwiren / @bcoca

**Comment by johanwiren** [comment]
Sorry, vacation mode here. Will have a look this evening.

**Comment by johanwiren** [comment]
#2642 looks OK to me. @bcoca, can you mark it ship_it if you approve of the change too.

**Comment by bcoca** [comment]
closing via merged #2642

**Comment by dcj** [comment]
This doesn't seemed fixed to me in ansible 2.1.2.0

```
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_9JcUwu/ansible_module_zfs.py", line 246, in <module>
    main()
  File "/tmp/ansible_9JcUwu/ansible_module_zfs.py", line 228, in main
    zfs = Zfs(module, name, properties)
  File "/tmp/ansible_9JcUwu/ansible_module_zfs.py", line 90, in __init__
    self.enhanced_sharing = self.check_enhanced_sharing()
  File "/tmp/ansible_9JcUwu/ansible_module_zfs.py", line 99, in check_enhanced_sharing
    if int(version) >= 34:
ValueError: invalid literal for int() with base 10: '-'
```

**Comment by xen0l** [comment]
@dcj, can you be more specific?

**Comment by dcj** [comment]
The target machine is SmartOS, build 20160929T025934Z

```
   - name: Create zfs datasets
      zfs:
        name: zones/images
        state: present
```

@xen0l, what else would you like to know?

**Comment by xen0l** [comment]
Well, it seems that fix is not yet available in 2.1.2.

**Comment by dcj** [comment]
@xen0l OK, thanks for the info.  I have hacked your fix info my module path for now.  When I read that the fix was merged in early August, I thought it must have been included in a late-September release...
